src/lem/circuit.rs

```diff
@@ -0,0 +1,904 @@
+
+                                // Add the implication constraint for the image
+                                implies_equal(
+                                    &mut cs.namespace(|| {
+                                        format!("implies equal for {}'s hash (LEMOP {:?})", $img, &op)
+                                    }),
+                                    &concrete_path,
+                                    $allocated_img.hash(),
+                                    &preallocated_img,
+                                )?;
+
+                                // For each component of the preimage, add implication constraints
+                                // for its tag and hash
+                                for (i, allocated_ptr) in $allocated_preimg.iter().enumerate() {
+                                    let name = $preimg[i].name();
+                                    let ptr_idx = 2 * i;
+                                    implies_equal(
+                                        &mut cs.namespace(|| {
+                                            format!("implies equal for {name}'s tag (LEMOP {:?}, pos {i})", &op)
+                                        }),
+                                        &concrete_path,
+                                        allocated_ptr.tag(),
+                                        &preallocated_preimg[ptr_idx], // tag index
+                                    )?;
+                                    implies_equal(
+                                        &mut cs.namespace(|| {
+                                            format!(
+                                                "implies equal for {name}'s hash (LEMOP {:?}, pos {i})",
+                                                &op
+                                            )
+                                        }),
+                                        &concrete_path,
+                                        allocated_ptr.hash(),
+                                        &preallocated_preimg[ptr_idx + 1], // hash index
+                                    )?;
+                                }
+                            };
+                        }
+                        macro_rules! hash_helper {
+                            ( $img: expr, $tag: expr, $preimg: expr, $slot: expr ) => {
+                                // Allocate image
+                                let allocated_img = LEM::allocate_ptr(
+                                    cs,
+                                    &LEM::zptr_from_mptr($img, &g.frame, g.store)?,
+                                    $img.name(),
+                                    &mut g.allocated_ptrs,
+                                )?;
+
+                                // Retrieve allocated preimage
+                                let allocated_preimg =
+                                    LEM::get_allocated_preimg($preimg, &g.allocated_ptrs)?;
```

Comment:
When/where are the preimages allocated? Can you point me to the code to help me trace the provenance?

Reply:
They must have been allocated when we do the constraining for the LEMOP that introduced them. For example:
```rust
let x: Terminal;
let y: Cons = hash2(x, x);
```
Then we have to constrain the `LEMOP::Hash2`, we need `x` to be available in the `allocated_ptrs`

Reply:
Preimages in LEM are variables that should be previously defined, either as an input or as result of a previous LEMOP. This means that necessarily they will be allocated. Otherwise, interpretation would fail, which comes before interpretation

Reply:
We can either check this statically with a function that takes a LEM and checks whether variables that appear an input of operations were previously bound. If everything is bound then the preimages should be allocated